### ISSUE TITLE
[BUG] Longhorn api-server PUT request rate

### ISSUE THREAD

**Issue by derekbit** [longhorn/longhorn#8114]
### Discussed in https://github.com/longhorn/longhorn/discussions/8112

<div type='discussions-op-text'>

<sup>Originally posted by **krohrsb** March  6, 2024</sup>
I guess my question comes down to checking the chattiness of longhorn PUT requests to the apiserver.

Specifically `PUT "/apis/longhorn.io/v1beta2/namespaces/longhorn-system/replicas/<some-pvc-id>"`

I have 52 volumes, each with 2 replicas, across 5 nodes. So a total of 104 replicas.

_**Prometheus denotes roughly 34 PUT requests per second by the longhorn group.**_

Compared to the next highest caller (2.5r/s), longhorn is about 1200% (1.2k%) higher :) 

So I am just wanting to confirm if this is expected, and if anything can be done to mitigate it without repercussions.

I realize its probably doing so for a reason, it just seemed high. But perhaps I just don't know better :) </div>

**Comment by krohrsb** [comment]
I've just sent another support bundle as it's continuing but now I'm also getting a high (but not as high as the replicas group) volume of PUTs to the node group.

**Comment by krohrsb** [comment]
Since both resource groups are now high it's pegging my cpu :/

![Screenshot_20240307-233405.png](https://github.com/longhorn/longhorn/assets/321544/f5a5e552-5bae-4328-a387-b994ddf965d7)

**Comment by krohrsb** [comment]
@ejweber I know you are working on some sort of alternative, I just thought I'd share:

Reviewing my journalctl logs, grepping for Updates to a specific replica, namely `,url:/apis/longhorn.io/v1beta2/namespaces/longhorn-system/replicas/pvc-52e929ab-41a3-4008-9294-a9338b1e9e2b-r-4c1dbdd9` yielded `24` operations across 1.5 minutes.

Now, the logs don't tell me what the update specifically was (I'm unsure if there are other normal PUT/Update operations against replicas). Just wanted to add just in case:

- There are other PUT's, which may be normal, alongside the LastHealthyAt updates, resulting in >  1/30s
- My system is so backlogged due to the overall issue & pressure its causing that it has queued requests that are getting sent out faster than they normally would, being in a perpetual queued state (I'm just not super familiar with this scenario)
- Its happening more often than 1/30s for some other reason.


I'm not trying to complicate matters, but figured more info is better than less. Do let me know if there is anything needed from me as well.

I appreciate all your efforts!

**Comment by krohrsb** [comment]
Just wanted to touch base real fast. I ended up converting my SQLite k3s into embedded etcd. This seems to have solved the bottleneck and thus my CPU pegged issue. Not longhorns fault overall but perhaps the slight delta in PUT throughput tipped my environment over the edge causing a cascading type of issue. What was interesting was that afterward I had like 500 Tps of put request from longhorn getting 409 conflict. Didn't stop after an hour or so so I restarted the longhorn manager pods and that seems to have solved that problem. I'm sure if there was some backlogged queue. Anyways sorry for being so chatty. Thanks again.

**Comment by ejweber** [comment]
Thanks for the additional information @krohrsb.

I think I can understand how the issue you are describing can be caused by the `lastHealthyAt` updates, though I am not sure we have enough data to fully verify it.

The way our controllers work, as soon as we push a replica update to the API server, longhorn-manager recognizes the replica was updated and requeues it for reconciliation. In the super simple example from my PR (https://github.com/longhorn/longhorn-manager/pull/2685), we update a replica's `lastHealthyAt` field once every 30 seconds. This causes a second reconciliation, which is very likely to occur within one second of the first. `lastHealthyAt` does not need to be modified in this case, so there is no additional update or reconciliation.

In your cluster, with many more replicas and potentially noticeable resource constraints (e.g. SQLite K3s backend), this flow may look quite different. Imagine we update the `lastHealthyAt` field for a replica. Maybe the update takes a bit of time due to API server constraints. Also, every replica on the node is updated at the same(ish) time. All of these updates trigger secondary reconciliations. 

Now, only five volumes (with their associated replicas) can be reconciled in parallel. If these reconciliations take a second, then the `lastHealthyAt` field for the next five replicas are out-of-date when they reconcile. So the volume controller pushes updated versions of these replicas (with updated `lastHealthyAt` times) to the API server. It seems possible for the whole system to get behind in such a way that it can never really catch up.

I tried it out with some extra logging and 10 volumes (30 replicas) in my same three node cluster.

Every time a replica is updated, its owning volume was immediately requeued.

```
time="2024-03-11T15:42:52Z" level=warning msg="Updated replica pvc-5e95215d-e45c-43b7-8869-92219df68a78-r-eb60f636 from resource version 58195276 to 58195532" func="controller.(*VolumeController).syncVolume.func2" file="volume_controller.go:442" controller=longhorn-volume node=eweber-v126-worker-9c1451b4-w6977
time="2024-03-11T15:42:52Z" level=warning msg="Enqueuing replica pvc-5e95215d-e45c-43b7-8869-92219df68a78-r-eb60f636 with resource version 58195532" func="controller.(*VolumeController).ResolveRefAndEnqueue" file="volume_controller.go:3514" controller=longhorn-volume node=eweber-v126-worker-9c1451b4-w6977
```

I saw an average of 38 replica PUTs every 30 seconds instead of the "expected" 30 (one for each replica).

![image](https://github.com/longhorn/longhorn/assets/24213029/d1caa434-8c07-43ec-9b10-b4155d4068be)

Then, I added a 1.5 second sleep every time a replica was updated (to try to trigger a worst case scenario in which `lastHealthyAt` would always need modification). I saw an average of 190 PUTs every 30 seconds instead of the "expected" 30.

![image](https://github.com/longhorn/longhorn/assets/24213029/5b8862bf-a20a-4754-8335-1a3e9fffa78b)

Applying the fix from https://github.com/longhorn/longhorn-manager/pull/2685 dropped the number of PUTs to zero (which we would expect under stable conditions), even with many volumes and the 1.5 second sleep.

![image](https://github.com/longhorn/longhorn/assets/24213029/ebfd2d1d-b78b-4e3d-a452-ba2cd1c03b78)

Long story short, I think @derekbit has correctly identified the problem, which hopefully https://github.com/longhorn/longhorn-manager/pull/2685 will fix.

**Comment by roger-ryao** [comment]
**Verified on master-head 20240319**

- longhorn `master-head` https://github.com/longhorn/longhorn/commit/42c075777c4b6d64cf6496caad232be5b2f0c4a5
- longhorn-manager `master-head` https://github.com/longhorn/longhorn-manager/commit/340c10a7256dd89b49e862b3edb88499bc5d62c3

**The test steps**
1. https://github.com/longhorn/longhorn-manager/pull/2432#issuecomment-1894675916

2. https://github.com/longhorn/longhorn-manager/pull/2685#issuecomment-1986537266
    1. Ensure that the upgrade correctly populates `replicaTransitionTimeMap`
    2. Confirm that the rebuild operation correctly populates `replicaTransitionTimeMap` and `lastHealthyAt` after replica deletion
    3. Validate that the rebuild operation correctly populates `replicaTransitionTimeMap` and `lastHealthyAt` after replica failure
    4. Ensure that replicas are not updated unnecessarily when the cluster is stable.
    
**Result Passed**

- [x] 1. If you wait any length of time, ReplicaA is never cleaned up.
- [x] 1. After deleting ReplicaB and waiting for the volume to rebuild from ReplicaA, the data is consistent.
- [x] 1. We observed that the `replicaTransitionTimeMap` is visible on `master-head`.
- [x] 1. no PUT requests to replica resources were observed.
- [x] 1. The rebuild operation correctly populates `replicaTransitionTimeMap` and `lastHealthyAt` after replica deletion
- [x] 1. The rebuild operation correctly populates `replicaTransitionTimeMap` and `lastHealthyAt` after replica failure

**Comment by krohrsb** [comment]
Just circling back as I have tried v1.6.1 and this has resolved my original reported issue. Much appreciated.

![lh](https://github.com/longhorn/longhorn/assets/321544/814b19f6-6494-4b33-89f5-5cd32c8ddc8e)

**Comment by serhiynovos** [comment]
Wonder when Rancher users will be able to try it out :). It still is showing 1.5.4 as latest version

**Comment by serhiynovos** [comment]
Hi. I guess it has to fix the problem what I'm facing, right ?

https://github.com/rancher/rke2/discussions/5680
